src/target/adi.c

```diff
@@ -589,17 +595,150 @@ static void adi_parse_adi_rom_table(adiv5_access_port_s *const ap, const target_
 		if (entry == 0)
 			break;
 
-		if (!(entry & ADIV5_ROM_ROMENTRY_PRESENT)) {
+		if (!(entry & ADI_ROM_ROMENTRY_PRESENT)) {
 			DEBUG_INFO("%s%" PRIu32 " Entry 0x%08" PRIx32 " -> Not present\n", indent, i, entry);
 			continue;
 		}
 
 		/* Probe recursively */
-		adi_ap_component_probe(ap, base_address + (entry & ADIV5_ROM_ROMENTRY_OFFSET), recursion_depth + 1U, i);
+		adi_ap_component_probe(ap, base_address + (entry & ADI_ROM_ROMENTRY_OFFSET), recursion_depth + 1U, i);
 	}
 	DEBUG_INFO("%sROM Table: END\n", indent);
 }
 
+static bool adi_reset_resources(adiv5_access_port_s *const ap, const target_addr64_t base_address)
+{
+	/* Read out power request ID register 0 and check if power control is actually implemented */
+	const uint8_t pridr0 = adi_mem_read32(ap, base_address + CORESIGHT_ROM_PRIDR0) & 0x3fU;
+	if ((pridr0 & CORESIGHT_ROM_PRIDR0_VERSION_MASK) != CORESIGHT_ROM_PRIDR0_VERSION_NOT_IMPL)
+		ap->flags |= ADIV6_DP_FLAGS_HAS_PWRCTRL;
+	/* Now try and perform a debug reset request */
+	if (pridr0 & CORESIGHT_ROM_PRIDR0_HAS_DBG_RESET_REQ) {
+		platform_timeout_s timeout;
+		platform_timeout_set(&timeout, 250);
+
+		adi_mem_write32(ap, base_address + CORESIGHT_ROM_DBGRSTRR, CORESIGHT_ROM_DBGRST_REQ);
+		/* While the reset request is in progress */
+		while (adi_mem_read32(ap, base_address + CORESIGHT_ROM_DBGRSTRR) & CORESIGHT_ROM_DBGRST_REQ) {
+			/* Check if it's been acknowledge, and if it has, deassert the request */
+			if (adi_mem_read32(ap, base_address + CORESIGHT_ROM_DBGRSTAR) & CORESIGHT_ROM_DBGRST_REQ)
+				adi_mem_write32(ap, base_address + CORESIGHT_ROM_DBGRSTRR, 0U);
+			/* Check if the reset has timed out */
+			if (platform_timeout_is_expired(&timeout)) {
+				DEBUG_WARN("adi: debug reset failed\n");
+				adi_mem_write32(ap, base_address + CORESIGHT_ROM_DBGRSTRR, 0U);
+				break;
+			}
+		}
+	}
+	/* Regardless of what happened, extract whether system reset is supported this way */
+	if (pridr0 & CORESIGHT_ROM_PRIDR0_HAS_SYS_RESET_REQ)
+		ap->flags |= ADIV6_DP_FLAGS_HAS_SYSRESETREQ;
+	return true;
+}
+
+static inline uint64_t adi_read_rom_class_9_entry(
```

Comment:
This function needs its nomenclature updated - `adi_read_coresight_rom_entry`?

Reply:
Oops, should be fixed :+1: